TypeError: unhashable type

**Describe the bug**
Call **setUp** fails on [fake_filesystem_unittest.py::_find_modules](https://github.com/pytest-dev/pyfakefs/blob/d17e190f35d52015043586af439d8e4c1f669404/pyfakefs/fake_filesystem_unittest.py#L798) when one of the modules has unhashable type. I don't know wether is a bad ([Undefined](https://pypi.org/project/undefined/)) module or not, but I think this is weird behavior.

```
self = <pyfakefs.fake_filesystem_unittest.Patcher object at 0x7fdc28337d50>

    def _find_modules(self) -> None:
        """Find and cache all modules that import file system modules.
        Later, `setUp()` will stub these with the fake file system
        modules.
        """
        module_names = list(self._fake_module_classes.keys()) + [PATH_MODULE]
        for name, module in list(sys.modules.items()):
            try:
                if (
                    self.use_cache
                    and module in self.CACHED_MODULES
                    or not inspect.ismodule(module)
                ):
                    continue
            except Exception:
                # workaround for some py (part of pytest) versions
                # where py.error has no __name__ attribute
                # see https://github.com/pytest-dev/py/issues/73
                # and any other exception triggered by inspect.ismodule
                if self.use_cache:
>                   self.__class__.CACHED_MODULES.add(module)
E                   TypeError: unhashable type: 'Undefined'
```

List of unhashable modules:
```python
>>> [i for i in sys.modules.items() if not isinstance(i[1], typing.Hashable)]
[('undefined', Undefined)]
```

**How To Reproduce**
1. Install suspicious unhashable module such as [Undefined](https://pypi.org/project/undefined/)
1. Run tests which uses pyfakefs

**Environment**
```
Linux-6.2.0-37-generic-x86_64-with-glibc2.35
Python 3.11.7 (main, Dec  8 2023, 18:56:58) [GCC 11.4.0]
pyfakefs 5.3.2
pytest 7.4.3
```

--- a/.github/workflows/testsuite.yml
+++ b/.github/workflows/testsuite.yml
@@ -114,7 +114,7 @@ jobs:
         run: |
           pip install -r requirements.txt
           pip install -U pytest==${{ matrix.pytest-version }}
-          pip install opentimelineio
+          pip install opentimelineio undefined
           pip install -e .
           if [[ '${{ matrix.pytest-version }}' == '4.0.2' ]]; then
              pip install -U attrs==19.1.0

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,6 +1,11 @@
 # pyfakefs Release Notes
 The released versions correspond to PyPI releases.
 
+## Unreleased
+
+### Fixes
+* fixes handling of unhashable modules which cannot be cached (see [#923](../../issues/923))
+
 ## [Version 5.3.2](https://pypi.python.org/pypi/pyfakefs/5.3.2) (2023-11-30)
 Bugfix release.
 

--- a/pyfakefs/fake_filesystem_unittest.py
+++ b/pyfakefs/fake_filesystem_unittest.py
@@ -805,7 +805,11 @@ class Patcher:
                 # see https://github.com/pytest-dev/py/issues/73
                 # and any other exception triggered by inspect.ismodule
                 if self.use_cache:
-                    self.__class__.CACHED_MODULES.add(module)
+                    try:
+                        self.__class__.CACHED_MODULES.add(module)
+                    except TypeError:
+                        # unhashable module - don't cache it
+                        pass
                 continue
             skipped = module in self.SKIPMODULES or any(
                 [sn.startswith(module.__name__) for sn in self._skip_names]

--- a/pyfakefs/pytest_tests/pytest_fixture_test.py
+++ b/pyfakefs/pytest_tests/pytest_fixture_test.py
@@ -13,6 +13,7 @@
 # Example for a test using a custom pytest fixture with an argument to Patcher
 
 import pytest
+import undefined
 
 import pyfakefs.pytest_tests.example as example
 from pyfakefs.fake_filesystem_unittest import Patcher
@@ -39,6 +40,12 @@ def test_example_file_passing_using_patcher():
     with Patcher(modules_to_reload=[example]) as patcher:
         patcher.fs.create_file(example.EXAMPLE_FILE, contents="stuff here")
         check_that_example_file_is_in_fake_fs()
+
+
+def test_undefined(fs):
+    # regression test for #923
+    with pytest.raises(NotImplementedError):
+        print(undefined)
 
 
 def check_that_example_file_is_in_fake_fs():

